File: backend/models/quiz.py
from typing import TYPE_CHECKING

from sqlalchemy import (
    Column,
    Integer,
    ForeignKey,
    String,
    DateTime,
    Boolean,
    ARRAY,
)
import enum

from sqlalchemy.orm import relationship

from sqlalchemy.ext.hybrid import hybrid_property
from sqlalchemy.sql.sqltypes import JSON
from .association_tables import (
    group_quiz_association_table,
    instructor_quiz_association_table,
)

from core.db import Base


class Quiz(Base):
    id = Column(Integer, primary_key=True)
    end_time = Column(DateTime)
    start_time = Column(DateTime)
    title = Column(String, nullable=True)
    description = Column(String, nullable=True)
    is_randomized = Column(Boolean, default=False)
    display_individual = Column(Boolean, default=False)
    group = relationship(
        "Group", secondary=group_quiz_association_table, backref="quiz"
    )
    instructor = relationship(
        "User", secondary=instructor_quiz_association_table, backref="quiz"
    )
    total_marks = Column(Integer, default=0)
    course_id = Column(Integer, ForeignKey("course.id", ondelete="cascade"))
    course = relationship("Course", backref="quiz")
    __tablename__ = "quiz"  # noqa


class QuizQuestion(Base):
    id = Column(Integer, primary_key=True)

    question_text = Column(String, nullable=True)
    question_image = Column(ARRAY(String), nullable=True)

    # if IMAGE_OPTIONS in combination with option_image is present then, we show all the image in option_image, and then show all the options present in options
    options = Column(JSON, nullable=False)
    marks = Column(Integer, default=0)
    # if IMAGE_Options present and answer == 0, then check answer_image
    answer = Column(ARRAY(Integer), nullable=True)

    quiz_id = Column(Integer, ForeignKey("quiz.id", ondelete="cascade"))
    quiz = relationship("Quiz", backref="question")

    __tablename__ = "quiz_question"  # noqa

    @hybrid_property
    def multiple(self):
        if len(self.answer) > 1:
            return True
        else:
            return False

File: mobile/lib/repository/models/quiz.dart
import 'package:equatable/equatable.dart';
import 'package:json_annotation/json_annotation.dart';
import 'package:sikshyalaya/repository/models/course.dart';
import 'package:sikshyalaya/repository/models/group.dart';
import 'package:sikshyalaya/repository/models/instructor.dart';

part 'quiz.g.dart';

@JsonSerializable()
class Quiz extends Equatable {
  final int? id;
  final Course? course;
  final String? end_time;
  final String? start_time;
  final String? title;
  final String? description;
  final bool? is_randomized;
  final bool? display_individual;
  final int? total_marks;
  final List<Group>? group;
  final List<Instructor>? instructor;

  const Quiz(
      {this.id,
      this.course,
      this.end_time,
      this.start_time,
      this.title,
      this.description,
      this.is_randomized,
      this.display_individual,
      this.total_marks,
      this.group,
      this.instructor});

  static const empty = Quiz(
      id: null,
      course: Course.empty,
      end_time: '',
      start_time: '',
      title: '',
      description: '',
      is_randomized: null,
      display_individual: null,
      total_marks: null,
      group: [Group.empty],
      instructor: [Instructor.empty]);

  factory Quiz.fromJson(Map<String, dynamic> json) => _$QuizFromJson(json);

  @override
  List<Object?> get props => [
        id,
        course,
        end_time,
        start_time,
        title,
        description,
        is_randomized,
        display_individual,
        total_marks,
        group,
        instructor
      ];
}
